spring-integration-zip/src/main/java/org/springframework/integration/zip/splitter/UnZipResultSplitter.java

```diff
@@ -21,32 +21,51 @@
 import java.util.Map;
 
 import org.apache.commons.io.FilenameUtils;
+
 import org.springframework.integration.file.FileHeaders;
 import org.springframework.integration.support.MessageBuilder;
 import org.springframework.integration.zip.ZipHeaders;
 import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHeaders;
 
 /**
  *
  * @author Gunnar Hillert
+ * @author Andriy Kryvtsun
  * @since 1.0
  *
  */
 public class UnZipResultSplitter {
 
-	public List<Message<Object>> splitUnzippedMap(Map<String, Object> unzippedEntries) {
-
-		final List<Message<Object>> messages = new ArrayList<Message<Object>>(unzippedEntries.size());
+	public List<Message<Object>> splitUnzippedMap(Message<Map<String, Object>> message) {
+		return createMessages(message.getPayload(), message.getHeaders());
+	}
 
+	private List<Message<Object>> createMessages(Map<String, Object> unzippedEntries, MessageHeaders headers) {
+		List<Message<Object>> messages = createList(unzippedEntries.size());
 		for (Map.Entry<String, Object> entry : unzippedEntries.entrySet()) {
-			final String path = FilenameUtils.getPath(entry.getKey());
-			final String filename = FilenameUtils.getName(entry.getKey());
-			final Message<Object> splitMessage = MessageBuilder.withPayload(entry.getValue())
-					.setHeader(FileHeaders.FILENAME, filename)
-					.setHeader(ZipHeaders.ZIP_ENTRY_PATH, path).build();
-			messages.add(splitMessage);
+			messages.add(createMessage(entry, headers));
 		}
 		return messages;
 	}
 
+	private static List<Message<Object>> createList(int initialCapacity) {
+		return new ArrayList<Message<Object>>(initialCapacity);
+	}
+
+	private static Message<Object> createMessage(Map.Entry<String, Object> entry, Map<String, ?> headers) {
+		return MessageBuilder.withPayload(entry.getValue())
+				.setHeader(FileHeaders.FILENAME, getName(entry))
+				.setHeader(ZipHeaders.ZIP_ENTRY_PATH, getPath(entry))
+				.copyHeaders(headers)
+				.build();
```

Comment:
Right, that is purpose: to return a `List<MessageBuilder<Object>>` instead for better performance in the wrapper, which creates messages anyway by its logic.

Since `Message` is immutable, adding any new header to it provides some overhead which we would like to avoid if we can change something in our code.

I may guess that you will start to say that it would be a breaking change, but we are here just only after an `M1` and such a change should be expectable.

From other hand we really can do that in the separate PR, of course. But for now I would like to know if we can do that at all. There might be need changes in some other place.

If you don't like to do that, that's fine, too. I'll take care about that myself after merging your PR.

Anyway I have a fully different idea in mind for the `ZipSplitter` solution. AS you noticed this `UnZipResultSplitter` does not have so much value and may be enough easy re-implement by end-user from zero.

So, the idea is based on the `AbstractMessageSplitter` implementation and so to get a gain from the direct framework code usage.